flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreSinkWriter.java

```diff
@@ -0,0 +1,163 @@
+        Map<Integer, RecordWriter> buckets = writers.get(partition);
+        if (buckets == null) {
+            buckets = new HashMap<>();
+            writers.put(partition.copy(), buckets);
+        }
+        return buckets.computeIfAbsent(
+                bucket,
+                k ->
+                        overwrite
+                                ? fileStoreWrite.createEmptyWriter(
+                                        partition, bucket, compactExecutor)
+                                : fileStoreWrite.createWriter(partition, bucket, compactExecutor));
+    }
+
+    @Override
+    public void write(RowData rowData, Context context) throws IOException {
+        RowKind rowKind = rowData.getRowKind();
+        SinkRecord record = recordConverter.convert(rowData);
+        RecordWriter writer = getWriter(record.partition(), record.bucket());
+        try {
+            writeToFileStore(writer, record);
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+        rowData.setRowKind(rowKind);
+    }
+
+    private void writeToFileStore(RecordWriter writer, SinkRecord record) throws Exception {
+        switch (record.rowKind()) {
+            case INSERT:
+            case UPDATE_AFTER:
+                if (record.key().getArity() == 0) {
+                    writer.write(ValueKind.ADD, record.row(), GenericRowData.of(1));
+                } else {
+                    writer.write(ValueKind.ADD, record.key(), record.row());
+                }
+                break;
+            case UPDATE_BEFORE:
+            case DELETE:
+                if (record.key().getArity() == 0) {
+                    writer.write(ValueKind.ADD, record.row(), GenericRowData.of(-1));
+                } else {
+                    writer.write(ValueKind.DELETE, record.key(), record.row());
+                }
+                break;
+        }
+    }
+
+    @Override
+    public List<LocalCommittable> prepareCommit(boolean flush) throws IOException {
+        try {
+            return prepareCommit();
+        } catch (Exception e) {
+            throw new IOException(e);
+        }
+    }
+
+    private List<LocalCommittable> prepareCommit() throws Exception {
+        List<LocalCommittable> committables = new ArrayList<>();
+        for (BinaryRowData partition : writers.keySet()) {
+            Map<Integer, RecordWriter> buckets = writers.get(partition);
+            for (Integer bucket : buckets.keySet()) {
+                RecordWriter writer = buckets.get(bucket);
+                LocalCommittable committable =
+                        new LocalCommittable(partition, bucket, writer.prepareCommit());
+                committables.add(committable);
+
+                // clear if no update
+                if (committable.increment().newFiles().isEmpty()) {
+                    closeWriter(writer);
+                    buckets.remove(bucket);
+                }
```

Comment:
Why do we do this? We might get updates in the following checkpoint interval.

Reply:
Next checkpoint should read from latest snapshot.

Reply:
This is because we must need some mechanism to clear writers, such as yesterday's partition that no longer needs to be written.

Reply:
Add comments to point out this aspect.